### PR TITLE
[CHK-1524] transactionId wrong format in close payment request

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Fix wrong transaction id serialization in close payment

#### List of Changes

use <UUID>.value() method in place of raw toString()

#### Motivation and Context

Wrong close payment request format

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by giovanniberti — APPROVED** [review]

**Review by giovanniberti — APPROVED** [review]

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 1]
```diff
@@ -286,6 +286,20 @@ fun wasAuthorizationRequested(tx: BaseTransaction): Boolean =
...
+      tx
+        .transactionAtPreviousState()
+        .map { txAtPreviousStep ->
+          txAtPreviousStep.isRight && wasAuthorized(txAtPreviousStep.get())
+        }
+        .orElse(false)
+    is BaseTransactionExpired -> wasAuthorized(tx.transactionAtPreviousState)
+    else -> false
+  }
+
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:302]: It isn't the perfect match of the !wasAuthorizationDenied.

```
fun wasAuthorizationDenied(tx: BaseTransaction): Boolean =
  when (tx) {
    is BaseTransactionWithCompletedAuthorization -> !tx.wasTransactionAuthorized()
    is TransactionWithClosureError ->
      tx
        .transactionAtPreviousState()
        .map { txAtPreviousStep ->
          txAtPreviousStep.isRight && wasAuthorizationDenied(txAtPreviousStep.get())
        }
        .orElse(false)
    is BaseTransactionExpired -> wasAuthorizationDenied(tx.transactionAtPreviousState)
    else -> false
  }
```
The transaction could go to `CLOSURE_ERROR` state coming from CANCELLATION_REQUESTED (by user) or from AUTHORIZATION_COMPLETED. So this method could return `false` either if the authorization was really denied (good case for us) or if the transaction was canceled by user (bad case for us). So I thought that an explicit method was better that the denied of this one.

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 2]
```diff
@@ -205,6 +199,15 @@ class NodeService(
     return nodeClient.closePayment(closePaymentRequest.awaitSingle()).awaitSingle()
   }
 
+  private fun evaluateTransactionStatus(it: BaseTransaction): String =
+    if (wasAuthorized(it)) {
+      "Autorizzato"
+    } else if (wasAuthorizationDenied(it) && wasAuthorizationRequested(it)) {
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:205]: Same as before

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 3]
```diff
@@ -205,6 +199,15 @@ class NodeService(
     return nodeClient.closePayment(closePaymentRequest.awaitSingle()).awaitSingle()
   }
 
+  private fun evaluateTransactionStatus(it: BaseTransaction): String =
```
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:202]: I choose `getTransactionDetailsStatus`

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:302]: I see your point.
Based on the state diagram we can have basically three scenarios:

1. PGS authorized the transaction -> AuthorizationResultDto == OK
2. PGS denied authorization for the transaction -> AuthorizationResultDto == KO
3. No authorization is performed at all (transaction cancelled by user) -> here we can associate an AuthorizationResultDto == null

So the authorization outcome result can be see as an optional value based on the transaction flow.

Maybe we can handle this with the following method

```
fun getAuthorizationOutcome(tx: BaseTransaction): AuthorizationResultDto? =
    when (tx) {
        is BaseTransactionWithCompletedAuthorization -> tx.transactionAuthorizationCompletedData.authorizationResultDto
        is TransactionWithClosureError ->
            tx
                .transactionAtPreviousState()
                .map { txAtPreviousStep ->
                    txAtPreviousStep.fold({ null }, { it.transactionAuthorizationCompletedData.authorizationResultDto })
                }
                .orElse(null)

        is BaseTransactionExpired -> getAuthorizationOutcome(tx.transactionAtPreviousState)
        else -> null
    }
```

and then refactor the wasAuthorizationDenied as:

```
fun wasAuthorizationDenied(tx: BaseTransaction): Boolean =
    getAuthorizationOutcome(tx) == AuthorizationResultDto.KO
```

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:205]: Using the `getAuthorizationOutcome` method the whole function can be refactored as: 

```
private fun getTransactionDetailsStatus(it: BaseTransaction): String =
        when (getAuthorizationOutcome(it)) {
            AuthorizationResultDto.OK -> "Autorizzato"
            AuthorizationResultDto.KO -> "Rifiutato"
            else -> "Annullato"
        }

```

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt]: Did it!

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Did it!

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 4]
```diff
@@ -23,6 +24,9 @@ import org.springframework.stereotype.Service
 import reactor.core.publisher.Mono
 
 const val TIPO_VERSAMENTO_CP = "CP"
+const val TRANSACTION_DETAILS_STATUS_CANCELED = "Annullato"
+const val TRANSACTION_DETAILS_STATUS_AUTHORIZED = "Autorizzato"
+const val TRANSACTION_DETAILS_STATUS_DENIED = "Rifiutato"
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:29]: Refactor to enum?

Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Did it!

**Review by pietro-tota — APPROVED** [review]
Well done 🚀  just a comment

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**